### ISSUE TITLE
SQLite Manager and new Firefox

### ISSUE THREAD

**Issue by kcranston** [datacarpentry/sql-ecology-lesson#199]
Amy Hodge points out on [this thread](http://lists.software-carpentry.org/pipermail/discuss/2017-November/005585.html) that the new release of Firefox does not support the SQLite Manager plugin. Couple of things we should do:

* (short term): update the setup instructions letting instructors / learners know which version of Firefox they need to be able to run the plugin
* (long term): explore other options. In the discuss thread, @fmichonneau suggests [Falcon](https://plot.ly/database-connectors/) and [sqlitebrowser](https://github.com/sqlitebrowser/sqlitebrowser) as possibilities.

**Comment by liz-is** [comment]
Would the Firefox [Extended Support Release](https://www.mozilla.org/en-US/firefox/organizations/) be useful here in the short term? It's intended to provide extended support for Java-based plugins, and may still support SQLite Manager.

**Comment by twitwi** [comment]
ESR are probably ok in the short term, up to june 2018: https://www.mozilla.org/en-US/firefox/organizations/faq/

**Comment by swaldman3** [comment]
This is correct - the ESR will allow old extensions, inclding SQLite Mgr. However, this is *only* a short-term option, as it reaches end of life in mid-2018.
Also: Firefox does not make it easy to have multiple versions installed, and I think we would get a lot of pushback if we asked learners to "downgrade" their browser version for the sake of the software that we want them to use.

**Comment by naupaka** [comment]
As an alternative to Firefox, there's [DBeaver](https://dbeaver.jkiss.org/). Apache license, works on all platforms. Also will work with many different types of database engines, not just SQLite.

**Comment by kcranston** [comment]
It looks like there are some alternatives out there (and perhaps some better UI than the current SQLIte Manager plugin!). What we need now is some testing. Anyone interested in participating? Maybe we could pick a day, have 1-2 people run through the lesson with each tool, and report back on the experience. 

If you are interested, leave a comment below.

**Comment by naupaka** [comment]
Happy to go through the lesson with DBeaver. I've been impressed with it for my own work.

**Comment by tpoisot** [comment]
I use [SQLite browser](http://sqlitebrowser.org/), but DBeaver looks nice.

**Comment by tpoisot** [comment]
And of course, there is a [SQLite backend](https://github.com/brownan/sqlite3-kernel) for Jupyter.

**Comment by PeterSmyth12** [comment]
DB Browser for SQLite from http://sqlitebrowser.org/ which a few people have mentioned seems to have all the required functionality, it even has a small graphing pane which seems like overkill to me.

I have just written the SQL lesson for the forthcoming Social Sciences curriculum, with several screenshots of the plugin and a tour of the facilities. 

I feel I will have to re-write it now, I could just leave out the details of the environment in use, but I would rather not.  For example creating a table using the GUI is likely to be slightly different depending on choice.

DBeaver which I have only just looked at would seem to have the advantage that it could be used with different database types, so presumably if anyone developed a lesson for MySQL or MariaDB the same interface could be used. 
On the downside (I'm speculating, as I haven' actually tried it) it could be more complex to use which wouldn't be good for people new to SQL in the first place.

As we expect the attendees to install the necessary s/w in advance I don't think we can go for any complicated solutions.

I will try out the Social Sciences SQL lesson with DB Brower for SQLite and report back.

**Comment by mkweskin** [comment]
Hi all- I just completed the Carpentries Instructor Training yesterday and used the "SQL for Ecology: Basic Queries" as the episode for my teaching exercises. This issue with Firefox dropping support for the SQLite Manager plugin came up during my training.

Today I tried both DBeaver and "DB Browser for SQLite" on my Mac with the beginning of the SQL for Ecology lesson and concur with @PeterSmyth12 that the DB Browser seems like the way to go. I found DB Browser very intuitive and was quickly able to create a new SQLite DB and import the CSVs. It looks like it has the about the same functionality as the Firefox plugin and I think the interface is easier (and the error messages less scary for novices). Full disclosure: I used an early version DB Browser years ago, so I wasn't a complete novice with it.

DBeaver is clearly extremely powerful, and I might check it out to work with some of my MySQL DB. However, I found it quite frustrating to figure how to create a new SQLite database and import the CSV files.

**Comment by naupaka** [comment]
Having now played with both, I agree with @mkweskin about it being quite challenging to make a new SQLite database from CSV files in DBeaver and that is probably reason enough not to use it for an intro workshop...

**Comment by PeterSmyth12** [comment]
Both DBeaver and DB Browser seem to have problems with importing csv files. 

There are instruction for DBeaver [here](https://github.com/serge-rider/dbeaver/issues/722)  ( a six point plan about half way down). There is also an outstanding request outlined here.

![dbeaver_csv](https://user-images.githubusercontent.com/25324961/32942292-825b691a-cb80-11e7-8da6-3e19c3b1b69f.png)
 
The version I downloaded is 4.2.5 so I guess this will be the next release.

Asking for the ability to change the datatypes on the load is in keeping with what you would expect and is/was available with the Firefox plugin. Currently I cannot get it to let me change the datatypes even after I have imported the file (as all TEXT fields) - The properties screen won't let me change anything.

DB Browser also doesn't give the option to change the datatypes on load, but at least it does allow me to modify the table structure afterwards. 

Another slight issue I have with DB Browser is that it seems to treat the whole session as a transaction, if you don't say you want to save the database at  the end you lose all of your changes.

What both have over the Firefox plugin though is the ability to save and load files of SQL and execute multiple statements at a time.

**Comment by mkweskin** [comment]
I checked out another open source, cross platform tool I found: [SQLiteStudio](https://sqlitestudio.pl).

- CSV import is straight forward, but you need to set the datatypes after the import
- Other than making structural changes to a table, you don't need to save to keep changes.
- There's an annoying GUI issue on my Mac where the file browser doesn't respond correctly when you create a new DB or import CSV so you have to type the full path in a text box . DBeaver had a similar issue with the file browser on my computer.

**Comment by orchid00** [comment]
Question: what do we do in the mean time, while deciding the next tool?

**Comment by tpoisot** [comment]
Let me summarize
 
| Tool             | Installation | Multi-platform | CSV import | SQLite only |
|:-----------------|:-------------|:---------------|:-----------|:------------|
| [SQLiteStudio]   | standalone   | yes            | **yes**    | **yes**     |
| [sqlite3-kernel] | Jupyter      | yes            | no         | **yes**     |
| [DBBrowser]      | standalone   | yes            | **yes**    | **yes**     |
| [DBeaver]        | standalone   | yes            | **yes**    | no          |

[SQLiteStudio]: https://sqlitestudio.pl/
[sqlite3-kernel]: https://github.com/brownan/sqlite3-kernel
[DBBrowser]: http://sqlitebrowser.org/
[DBeaver]: https://dbeaver.jkiss.org/

**Comment by danielsmaxwell** [comment]
I've started working with SQLiteStudio and like the interface.  It appears to be a rather capable tool.

**Comment by remram44** [comment]
Those are all opensource and multiplatform. sqlite3-kernel really stands apart (and is definitely harder to setup, though could be hosted via JupyterHub), I guess we'll need to look into the individual features...
<table>
<tr><th>Tool</th><th>License</th></tr>
<tr><td>SQLiteStudio</td><td>GPLv2</td></tr>
<tr><td>sqlite3-kernel</td><td>BSD-3</td></tr>
<tr><td>DBBrowser for SQLite</td><td>Mozilla & GPL</td></tr>
<tr><td>DBeaver</td><td>Apache 2.0</td></tr>
</table>

**Comment by PeterSmyth12** [comment]
For the SQL lesson in the Social Sciences curriculum  I have decided to use DB Browser. Perhaps not perfect, but straightforward to install and use. 

I have attached word documents covering the installation, launching and creating a table.
[Installing db browser.docx](https://github.com/datacarpentry/sql-ecology-lesson/files/1534820/Installing.db.browser.docx)
[Launching DB Browser.docx](https://github.com/datacarpentry/sql-ecology-lesson/files/1534821/Launching.DB.Browser.docx)
[Using the DB Browser application to create tables.docx](https://github.com/datacarpentry/sql-ecology-lesson/files/1534822/Using.the.DB.Browser.application.to.create.tables.docx)

All SQL runs as expected, although you need to be quite particular about including the ';' at the end of the statements. (even if you are just selecting a single statement to execute from the pane, if any of the previous statements are not terminated with a ';' it gets upset).

Although several of us have moaned about having to change data-types after creating a table. The modify table window does have the nice facility whereby as you change the options in the upper pane, you can see the DDL statement being modified in the lower pane. I see this as a great aid for teaching.

**Comment by justinclift** [comment]
@PeterSmyth12 As one of the DB Browser for SQLite developers, that sounds pretty awesome. :smile:

Not sure if it's useful for your audience, but we've rewritten large parts of our CSV import code since our last major release.  It now imports significantly faster (at least 3x as fast in most cases), and it auto-detects the closest matching data type on the way through.

Naturally though, you can still (manually) change the data types afterwards if you want. :smile:

We're also fairly responsive to bug reports, feature requests, and that kind of thing too:

&nbsp; &nbsp; https://github.com/sqlitebrowser/sqlitebrowser/issues

**Comment by remram44** [comment]
sqlitebrowser ("DBBrowser for SQLite") seems very good. The interface is not too cluttered, it is easy to import and export CSV, viewing/changing the data is intuitive. The schema editor is available under a button, it seems complete enough (though I was confused for a moment trying to create an index -- that is a separate button from the main view). There are obvious commit/rollback buttons, and plotting functionality that I guess is useful in some cases. Mac download is 14MB.

<details>
<summary>screenshots</summary>
<img width="320" alt="screen shot 2017-12-09 at 20 44 24" src="https://user-images.githubusercontent.com/426784/33801265-ec7d2040-dd23-11e7-9266-578baf42cf02.png">
<img width="320" src="https://user-images.githubusercontent.com/426784/33801266-f20935a8-dd23-11e7-8b13-fd97f0f29094.png">
<img width="320" src="https://user-images.githubusercontent.com/426784/33801267-f5b285ec-dd23-11e7-9dcf-c30a097d9ed0.png">
<img width="320" src="https://user-images.githubusercontent.com/426784/33801268-f8680b22-dd23-11e7-8a1f-ff1aa7d3ad73.png">
</details>

SQLiteStudio is also very complete. The interface is a bit more cluttered (multiple toolbars with 10+ icon-only buttons; I'm also unsure about the multi-window feature, with the navbar at the bottom). It seems to be putting more emphasis on editing the schema (default view) but the data is one-click away in another tab. Can open multiple databases at the same time. Has a user-defined-function editor. Mac download is 25MB.

<details>
<summary>screenshots</summary>
<img width="320" src="https://user-images.githubusercontent.com/426784/33801269-050b1676-dd24-11e7-9ce4-679f0c3ce961.png">
<img width="320" src="https://user-images.githubusercontent.com/426784/33801270-087d1b92-dd24-11e7-9f7b-b252ff75aba4.png">
<img width="320" src="https://user-images.githubusercontent.com/426784/33801271-0b687b12-dd24-11e7-9e06-820f9b676256.png">
</details>

DBeaver is Java, based on Eclipse. Takes a while to load, goes through scary assistant to create a "connection", asks to download SQLite3 JDBC driver. Interface is based on Eclipse, with multiple panes such as "projects", "database navigator". Seems to be aimed at power users more than being a GUI for beginners. Unfortunately I couldn't get past the JDBC driver download because that kept failing. Mac download is 43MB.

Falcon is a multi-backend (like DBeaver) SQL client by plotly. Making plot seems to be the primary application of this tool. I could not see a way to edit the data nor to view or edit the schema of a database, I couldn't display the content of a table without typing the SELECT as SQL. All it does is plot, send data to plotly, or export to CSV. Mac download is 90MB.

<details>
<summary>screenshot</summary>
<img width="523" src="https://user-images.githubusercontent.com/426784/33801272-18a9c628-dd24-11e7-8b48-be0383ed2919.png">
</details>

My vote so far is SQLitebrowser.

**Comment by justinclift** [comment]
As a data point, we used the abbreviation "DB4S" for "DB Browser for SQLite".  Much easier to type. 😁

**Comment by amyehodge** [comment]
I have submitted a PR that addresses this in the short-term, i.e. notifies people of the problem and changes the download links to the Firefox ESR page.

**Comment by ChristinaLK** [comment]
one data point: we used db browser for sqlite this week and it worked great.

**Comment by justinclift** [comment]
Possibly useful, one of our team members is adding [bar charts](https://github.com/sqlitebrowser/sqlitebrowser/pull/1302) to our "Plot" panel. eg:

#### "Stacked bars" checked and "Show legend" checked
![imagen](https://user-images.githubusercontent.com/13812910/35196026-e570a2d4-fecc-11e7-9fa4-3702a03d4009.png)

#### "Stacked bars" unchecked (grouped) and "Show legend" unchecked (defaults)
![imagen](https://user-images.githubusercontent.com/13812910/35196033-07b22976-fecd-11e7-83f1-6a35c2dda02d.png)

Kind of thinking that might be useful both in classes, and in general use. :smile:

The feature code itself is working and ready for review (not yet started).  Hopefully that'll be done in the next few days, after which it'll be available in our nightly builds (and next release onwards).

In the meantime, if anyone wants to give it a whirl we've made Windows (32 and 64-bit) builds available for testing:

* https://nightlies.sqlitebrowser.org/win32-onceoffs/DB%20Browser%20for%20SQLite-bar_charts_v2-win32.exe
* https://nightlies.sqlitebrowser.org/win64-onceoffs/DB%20Browser%20for%20SQLite-bar_charts_v2-win64.exe

**Comment by remram44** [comment]
@justinclift Plotting is not really the focus of the SQL lessons, but it's good to see that DB4S is so actively developed.

Maintainers: any objections to using DB4S? If people are already using it for teaching, we should go ahead and officially request a PR to update the lesson. Thanks! I can take a crack at it, but not in the next few days.

**Comment by justinclift** [comment]
@remram44 No worries at all. :smile:

**Comment by dannguyen** [comment]
OK I just stumbled upon this thread when searching for the state of SQLite3 in Firefox and other browsers (I had heard that SQLite Manager had stopped working, from other data journalism editors). I just want to say that I was pleasantly surprised to discover that DB Browser for SQLite, which is what I've used for a few years (with great satisfaction) for teaching, apparently has visualization functions! 

Very cool. I like DB4S because of how good it is as just a SQLite client. I teach that visualization (and many other data processes) should involve porting data to other tools, but just wanted to echo the sentiment of how cool it is that DB4S is so actively and well-maintained. It's one of my favorite and most personally-used open source projects, and data journalism orgs should be giving the DB4S team awards :)
